### PR TITLE
Documented import/export trade-offs (resolves #875)

### LINKED ISSUE #875: Document import/export trade-offs
Mention that 
- per-sample files should be imported on a worker
- it may be more efficient to not import large files that are only used once, but rather download them directly
- it may matter whether an external file in S3 is in the same region as the job store

Point out that toil-scripts/lib has a common functionality for downloading URLs.

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-343)
┆Issue Number: TOIL-343

### PR DESCRIPTION
**Pull request by JakeNarkizian**
Resolves #875

### CODE REVIEW

File: docs/developing.rst [hunk 1]
```diff
@@ -571,6 +571,13 @@ Example::
 
             toil.exportFile(outputFileID, 'file:///some/other/local/path')
 
+
+There are some trade-offs to file staging. For example it may be more efficient to directly download
+a large file if it is used once rather than importing it. Note that toil-scripts/lib contains common
```
Comment (hannes-ucsc) [inline docs/developing.rst]: Add link.

Comment (hannes-ucsc) [inline docs/developing.rst]: Why could it be more efficient? Explain.

File: docs/developing.rst [hunk 2]
```diff
@@ -571,6 +571,13 @@ Example::
...
+functionality for downloading URLs. Further imports from S3 will be more efficient if the source is
```
Comment (hannes-ucsc) [inline docs/developing.rst]: Do you mean "Furthermore, "?

File: docs/developing.rst [hunk 3]
```diff
@@ -571,6 +571,13 @@ Example::
...
+also an S3 bucket in the same region. Finally, per-sample files should be imported on a worker not
```
Comment (hannes-ucsc) [inline docs/developing.rst]: same region as what?

File: docs/developing.rst [hunk 4]
```diff
@@ -571,6 +571,13 @@ Example::
...
+the leader.
```
Comment (hannes-ucsc) [inline docs/developing.rst]: And how would that be done?

And when is that not an option?

Comment (hannes-ucsc) [inline docs/developing.rst]: And why is that better?

File: docs/developing.rst (hunk 1 above)
Comment (JakeNarkizian) [inline docs/developing.rst]: Link to the file in the GH repo?

Comment (hannes-ucsc) [inline docs/developing.rst]: Yes.

File: docs/developing.rst [hunk 5]
```diff
@@ -571,6 +571,19 @@ Example::
 
             toil.exportFile(outputFileID, 'file:///some/other/local/path')
 
+
+There are some trade-offs to file staging. For example it may be more efficient to directly download
+a large file if it is used once. Importing will be more efficient if the file is needed several
+times during a workflow and the job store's storage platform supports certain optimizations. For
+example in S3 importing should be used if the source file is in a different region than the job
+store. Note that toil-scripts/lib contains common functionality for downloading URLs (see
+https://github.com/BD2KGenomics/toil-scripts/blob/master/src/toil_scripts/lib/urls.py) in cases
```
Comment (hannes-ucsc) [inline docs/developing.rst]: Learn how to link in Sphinx docs.

File: docs/developing.rst [hunk 6]
```diff
@@ -571,6 +571,19 @@ Example::
 
             toil.exportFile(outputFileID, 'file:///some/other/local/path')
 
+
+There are some trade-offs to file staging. For example it may be more efficient to directly download
```
Comment (hannes-ucsc) [inline docs/developing.rst]: comma after example

File: docs/developing.rst [hunk 7]
```diff
@@ -571,6 +571,19 @@ Example::
...
+a large file if it is used once. Importing will be more efficient if the file is needed several
```
Comment (hannes-ucsc) [inline docs/developing.rst]: `only` before `once`

Comment (hannes-ucsc) [inline docs/developing.rst]: `Importing it`

File: docs/developing.rst [hunk 8]
```diff
@@ -571,6 +571,19 @@ Example::
...
+times during a workflow and the job store's storage platform supports certain optimizations. For
```
Comment (hannes-ucsc) [inline docs/developing.rst]: certain? which ones?

File: docs/developing.rst [hunk 9]
```diff
@@ -571,6 +571,19 @@ Example::
...
+example in S3 importing should be used if the source file is in a different region than the job
```
Comment (hannes-ucsc) [inline docs/developing.rst]: comma after `For example`

File: docs/developing.rst (hunk 8 above)
Comment (hannes-ucsc) [inline docs/developing.rst]: NM

File: docs/developing.rst (hunk 9 above)
Comment (hannes-ucsc) [inline docs/developing.rst]: Explain why.

File: docs/developing.rst [hunk 10]
```diff
@@ -571,6 +571,19 @@ Example::
...
+store. Note that toil-scripts/lib contains common functionality for downloading URLs (see
+https://github.com/BD2KGenomics/toil-scripts/blob/master/src/toil_scripts/lib/urls.py) in cases
+where a file is used once. Also, per-sample files should be imported on a worker not the leader.
+This can be done inside job functions with :func:`job.FileStore.importFile`. Using this method will
```
Comment (hannes-ucsc) [inline docs/developing.rst]: `fileStore` and its a method so link should use `meth`, right?

File: docs/developing.rst [hunk 11]
```diff
@@ -571,6 +571,19 @@ Example::
...
+allow for performance gains due to parallelization. In contract The
```
Comment (hannes-ucsc) [inline docs/developing.rst]: What parallelization? It will also retry the import if it fails.

File: docs/developing.rst [hunk 12]
```diff
@@ -571,6 +571,19 @@ Example::
...
+:func:`toil.common.Toil.importFile` method described above is intended for files that are required
```
Comment (hannes-ucsc) [inline docs/developing.rst]: The part after `that` is convoluted. No one requires files to be on the leader. They're just there. And if you want to import them, you need to do it on the leader.

**Comment by hannes-ucsc** [comment]
No more spelling and grammar mistakes, please. I'm tired of pointing those out.

File: docs/developing.rst [hunk 13]
```diff
@@ -571,6 +571,19 @@ Example::
...
+to be local for the leader.
```
Comment (hannes-ucsc) [inline docs/developing.rst]: `local to` not `local for`

File: docs/developing.rst (hunk 10 above)
Comment (JakeNarkizian) [inline docs/developing.rst]: I don't see ```meth``` anywhere in this file. Other methods also use ```fund``` like 
```
e.g. :func:`toil.job.Job.addChild` 
and :func:`toil.job.Job.addFollowOn`. 
```
But I can change it to ```meth``` if those are technically incorrect

Comment (hannes-ucsc) [inline docs/developing.rst]: http://www.sphinx-doc.org/en/stable/domains.html#cross-referencing-python-objects

File: docs/developing.rst [hunk 14]
```diff
@@ -571,6 +571,25 @@ Example::
...
+
+There are some trade-offs to file staging that should be considered. It may be more efficient to
+directly download a file for cases in which the cost of the copy operation imposed by an import is
+unnecessary. For example, a download is better for large files that are used only once. While
+importing is more efficient if the file in question is needed several times, provided that the
+storage platform of the job store supports certain optimizations gained by the import. For example,
+Amazon's S3 platform, used in the AWS job store, is better optimized for reads and writes to files
+that are in the same region. Thus an import is justified if a source file is in a different region
+than that of the job store. Note that toil-scripts/lib contains common functionality for downloading
+URLs (see `https://github.com/BD2KGenomics/toil-scripts/blob/master/src/toil_scripts/lib/urls.py
```
Comment (hannes-ucsc) [inline docs/developing.rst]: Use the separate syntax, where the target URL is listed below the paragraph. See http://www.sphinx-doc.org/en/stable/rest.html#external-links.

Also, the text `toil-scripts/lib` should be what's clickable, not the URL itself. The reader should only see the URL when hovering over the link, or after clicking on it.

File: docs/developing.rst [hunk 15]
```diff
@@ -571,6 +571,25 @@ Example::
...
+<https://github.com/BD2KGenomics/toil-scripts/blob/master/src/toil_scripts/lib/urls.py>`_) in cases
+where a download is preferred to file staging.
+
+Files that are not needed on multiple jobs, per-sample files for example, should be imported on
+workers. This is done inside job functions with calls to the :meth:`job.FileStore.importFile`
```
Comment (hannes-ucsc) [inline docs/developing.rst]: Make sure this link actually works in the generated output.

File: docs/developing.rst [hunk 16]
```diff
@@ -571,6 +571,25 @@ Example::
 
             toil.exportFile(outputFileID, 'file:///some/other/local/path')
 
+
+There are some trade-offs to file staging that should be considered. It may be more efficient to
+directly download a file for cases in which the cost of the copy operation imposed by an import is
+unnecessary. For example, a download is better for large files that are used only once. While
```
Comment (hannes-ucsc) [inline docs/developing.rst]: add `direct` before download

File: docs/developing.rst [hunk 17]
```diff
@@ -571,6 +571,25 @@ Example::
...
+importing is more efficient if the file in question is needed several times, provided that the
```
Comment (hannes-ucsc) [inline docs/developing.rst]: the part after provided is awkward

File: docs/developing.rst [hunk 18]
```diff
@@ -571,6 +571,25 @@ Example::
...
+storage platform of the job store supports certain optimizations gained by the import. For example,
+Amazon's S3 platform, used in the AWS job store, is better optimized for reads and writes to files
+that are in the same region. Thus an import is justified if a source file is in a different region
```
Comment (hannes-ucsc) [inline docs/developing.rst]: same region as what?

File: docs/developing.rst [hunk 19]
```diff
@@ -571,6 +571,25 @@ Example::
...
+than that of the job store. Note that toil-scripts/lib contains common functionality for downloading
```
Comment (hannes-ucsc) [inline docs/developing.rst]: That's only true if the workflow is run on EC2.

File: docs/developing.rst [hunk 20]
```diff
@@ -571,6 +571,25 @@ Example::
...
+URLs (see `https://github.com/BD2KGenomics/toil-scripts/blob/master/src/toil_scripts/lib/urls.py
+<https://github.com/BD2KGenomics/toil-scripts/blob/master/src/toil_scripts/lib/urls.py>`_) in cases
+where a download is preferred to file staging.
+
+Files that are not needed on multiple jobs, per-sample files for example, should be imported on
```
Comment (hannes-ucsc) [inline docs/developing.rst]: needed by

File: docs/developing.rst [hunk 21]
```diff
@@ -571,6 +571,24 @@ Example::
 
             toil.exportFile(outputFileID, 'file:///some/other/local/path')
 
+
+There are some trade-offs to file staging that should be considered. It may be more efficient to
+directly download a file for cases in which the cost of the copy operation imposed by an import is
+unnecessary. For example, a direct download is better for large files that are used only once. While
+importing can be more efficient in cases where the file in question is needed several times. For
```
Comment (hannes-ucsc) [inline docs/developing.rst]: Not a complete sentence.

**Comment by JakeNarkizian** [comment]
Jenkins, test this please.

File: docs/developing.rst [hunk 22]
```diff
@@ -571,6 +571,24 @@ Example::
 
             toil.exportFile(outputFileID, 'file:///some/other/local/path')
 
+
+There are some trade-offs to file staging that should be considered. It may be more efficient to
+directly download a file for cases in which the cost of the copy operation imposed by an import is
+unnecessary. For example, a direct download is better for large files that are used only once.
+Importing can be more efficient in cases where the file in question is needed several times. For
+example, Amazon's S3 platform, used in the AWS job store, is better optimized for reads and writes
+to files that are in the same region as that of the job store. Thus an import is justified if a
```
Comment (hannes-ucsc) [inline docs/developing.rst]: Actually, it is optimized for access by EC2 instances running in the same region as the bucket.

File: docs/developing.rst [hunk 23]
```diff
@@ -571,6 +571,24 @@ Example::
...
+source file is in a different region than that of the job store and the work flow is running on an
```
Comment (hannes-ucsc) [inline docs/developing.rst]: `workflow`

File: docs/developing.rst [hunk 24]
```diff
@@ -571,6 +571,24 @@ Example::
...
+AWS EC2 instance. Note that `toil-scripts/lib <https://github.com/BD2KGenomics/toil-scripts/blob/
+master/src/toil_scripts/lib/urls.py>`_ contains common functionality for downloading URLs in cases
```
Comment (hannes-ucsc) [inline docs/developing.rst]: As I already mentioned in a previous line note, use the linking syntax that places the destination URL below the paragraph.

File: docs/developing.rst [hunk 25]
```diff
@@ -571,6 +571,24 @@ Example::
...
 
+
+There are some trade-offs to file staging that should be considered. It may be more efficient to
+directly download a file for cases in which the cost of the copy operation imposed by an import is
+unnecessary. For example, a direct download is better for large files that are used only once.
+Importing can be more efficient in cases where the file in question is needed several times. For
+example, Amazon's S3 platform, used in the AWS job store, is better optimized for reads and writes
+to files that are in the same region as that of the job store. Thus an import is justified if a
+source file is in a different region than that of the job store and the work flow is running on an
+AWS EC2 instance. Note that `toil-scripts/lib <https://github.com/BD2KGenomics/toil-scripts/blob/
```
Comment (hannes-ucsc) [inline docs/developing.rst]: Could be more than one instance.

File: docs/developing.rst [hunk 26]
```diff
@@ -571,6 +571,24 @@ Example::
...
+master/src/toil_scripts/lib/urls.py>`_ contains common functionality for downloading URLs in cases
+where a download is preferred to file staging.
+
+Files that are not needed by multiple jobs, per-sample files for example, should be imported on
```
Comment (hannes-ucsc) [inline docs/developing.rst]: `Files that are accessed exactly once during a workflow, …` and delete the `per-sample` part.

File: docs/developing.rst [hunk 27]
```diff
@@ -571,6 +571,25 @@ Example::
...
+AWS EC2 instances. Note that `toil-scripts/lib`_ contains common functionality for downloading URLs in cases
+where a download is preferred to file staging.
+
+.. _toil-scripts/lib: https://github.com/BD2KGenomics/toil-scripts/blob/master/src/toil_scripts/lib/urls.py
+
+
+Files that are accessed exactly once during a workflow, should be imported on workers. This is done
+inside job functions with calls to the :meth:`toil.job.Job.FileStore.importFile` method. Importing
+files on a worker where possible will allow for performance gains due to the parallelization of
+concurrently running jobs as well as the added benefit of retries on import failures.
```
Comment (hannes-ucsc) [inline docs/developing.rst]: Still not happy with this. I think we should mention the different cases first and then list the pros and cons for each one. The cases are: 

1. Import
    
    1a. Import on leader
    
    1b. Import on worker

2. Direct download on worker

Note how 1a and 1b are subitems of 1. They share certain pros and cons.

Be sure to mention that files local to the leader must be imported on the leader.

File: docs/developing.rst [hunk 28]
```diff
@@ -571,6 +571,28 @@ Example::
 
             toil.exportFile(outputFileID, 'file:///some/other/local/path')
 
+
+Files can be staged in different ways. A file can be imported on either the leader or worker, or it
```
Comment (hannes-ucsc) [inline docs/developing.rst]: a worker

Comment (hannes-ucsc) [inline docs/developing.rst]: imported into the job store

File: docs/developing.rst [hunk 29]
```diff
@@ -571,6 +571,28 @@ Example::
...
+can be directly downloaded on the worker. Each of these approaches have their own advantages that
```
Comment (hannes-ucsc) [inline docs/developing.rst]: to a worker

File: docs/developing.rst [hunk 30]
```diff
@@ -571,6 +571,28 @@ Example::
...
+should be considered.
+
+Importing is preferred for cases in which the file in question is needed several times. For
+example, Amazon's S3 platform, used in the AWS job store, is better optimized for reads and writes
```
Comment (hannes-ucsc) [inline docs/developing.rst]: S3 is a service

Comment (hannes-ucsc) [inline docs/developing.rst]: Toil's AWS job store implementation

Comment (hannes-ucsc) [inline docs/developing.rst]: delete `better`

File: docs/developing.rst [hunk 31]
```diff
@@ -571,6 +571,28 @@ Example::
...
+to buckets from an EC2 instance in the same region. Thus an import is justified if a source file is
```
Comment (hannes-ucsc) [inline docs/developing.rst]: same region as what?

**Comment by hannes-ucsc** [comment]
It's not clear. We'll have to revisit this at a later time.

**Pull request closed without merging**